Explicitely mention that the PyTorch easyblock needs updating when failing for this reason

Flamefire opened the pull request:
(created using `eb --new-pr`)

See https://github.com/easybuilders/easybuild-easyconfigs/pull/19666#issuecomment-1991295933 for the motivation

casparvl commented on easybuild/easyblocks/p/pytorch.py:
```diff
@@ -477,9 +477,16 @@ def test_step(self):
+            uncounted_test_suites = all_failed_test_suites - failed_test_suites
+            if uncounted_test_suites:
+                failure_report.append('Could not count failed tests for the following test suites/files:')
+                failure_report.extend(sorted(uncounted_test_suites))
```
I think (after listing the uncounted test suites) we could add a more explicit explanation and instruction. Something like:

"This will always result in a failure of the test step, because we do not know how many tests have failed, and thus cannot compare to the `max_failed_tests` set in the EasyConfig. Please investigate the cause of failures in these test suite. If you are _sure_ they are harmless, you can consider running with `--ignore-test-failures` to make the test step pass."

Flamefire replied:
I did think about adding something like "This will always result in a failure of the test step" but decided for a more clear `raise EasyBuildError('Failing because the PyTorch EasyBlock needs updating!\n' + msg)`

Especially due to the 500 char limit printed by EB I wanted to keep the important information very short. (The full/detailed message is appended)

However I like the instructions for users. So maybe below use: `raise EasyBuildError('Failing because the test accounting in the PyTorch EasyBlock needs updating!\nYou can check the test failures (in the log) manually and if they are harmless, use --ignore-test-failures to make the test step pass.\n' + msg)`

And update & move the comment there (line 501/508) to "We can't be sure how many tests failed so comparing to max_failed_tests cannot reasonably be done"

Flamefire commented:
@casparvl I updated the EasyBlock after the discussion at https://github.com/easybuilders/easybuild-easyblocks/pull/3255#discussion_r1524517112

I also noticed and fixed 2 things:
- If we can't count failures we only failed if we counted ANY failure
- We can't count failures if a test aborted but the previous iteration showed a message that the EasyBlock needed updating

I moved the detection of "uncounted suites" into the parser function such that we can test it (manually).

Flamefire commented:
Can this be merged? Working on this easyblock again and want to avoid conflicts and potentially use the improvements from here

Flamefire commented:
Test report by @Flamefire
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _PyTorch-2.0.1-foss-2022atest-report.eb_ (partial log available at https://gist.github.com/Flamefire/9f058a380a9a4487d8cdc19246755e40)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
i7004 - Linux Rocky Linux 8.9 (Green Obsidian), x86_64, AMD EPYC 7702 64-Core Processor (zen2), Python 3.8.17
See https://gist.github.com/Flamefire/5d45465aabf55d4d84a543be0af94893 for a full test report.

akesandgren reviewed (approved):
LGTM

akesandgren commented:
@Flamefire do you want to change anything else or should we merge as is?

The failed test report came in while I was reviewing :-)

Flamefire commented:
I added some further polishing, yes. It can be merged now. The test report was for an EasyConfig I made to run in ~40 minutes and fail some suites in different ways. It was meant to show the new output:

```
WARNING Found 3 individual tests that exited with an error: test_checkpointing_without_reentrant_detached_tensor_use_reentrant_False, test_excessive_thread_creation_warning, test_excessive_thread_creation_warning
Found 2 individual tests with failed assertions: test_fn_grad_linalg_det_singular_cpu_float64, test_memory_timeline

ERROR EasyBuild crashed with an error (at easybuild/base/exceptions.py:126 in __init__): Failing because not all failed tests could be determined. The test accounting in the PyTorch EasyBlock needs updating!
Missing: test_sparse, test_sparse_csr
You can check the test failures (in the log) manually and if they are harmless, use --ignore-test-failures to make the test step pass.
2 test failures, 3 test errors (out of 6437):
Failed tests (suites/files):
	profiler/test_memory_profiler (32 total tests, failures=1)
	distributed/_tensor/test_dtensor_ops (637 total tests, skipped=36, expected failures=407, unexpected successes=1)
	test_autograd (554 total tests, errors=1, skipped=24, expected failures=1)
	test_dataloader (164 total tests, errors=2, skipped=15)
	test_ops_gradients (1 failed, 1892 passed, 3055 skipped, 42 xfailed, 1 warning, 2 rerun)
Could not count failed tests for the following test suites/files:
	test_sparse (Did not run properly)
	test_sparse_csr (Did not run properly) (at easybuild/easyblocks/pytorch.py:582 in test_step)
```

I updated that slightly because the missing tests are those that failed to start at all. In the (current) log this shows up as:
```
Running test_sparse_csr ... [2025-02-21 11:02:19.117097]
Executing ['/data/Python/3.10.4-GCCcore-11.3.0/bin/python', '-bb', 'test_sparse_csr.py', '-v'] ... [2025-02-21 11:02:19.117297]

Traceback (most recent call last):
  File "/dev/shm//pytorch-v2.0.1/test/test_sparse_csr.py", line 23, in <module>
    from test_sparse import CUSPARSE_SPMM_COMPLEX128_SUPPORTED
  File "/dev/shm/pytorch-v2.0.1/test/test_sparse.py", line 29, in <module>
    raise RuntimeError("Intended failure")
RuntimeError: Intended failure
FINISHED PRINTING LOG FILE of test_sparse_csr (/dev/shm/pytorch-v2.0.1/test/test-reports/test_sparse_csr_qph3b9p4.log)
```

So without the last commit there is a discrepancy: It shows (in this case correctly) that the tests failed to run properly but also that the test parsing needs updating which is not the case here: It needs a patch to fix the test. Usually this specific issue is caused by us applying a patch from a previous version that breaks the current version.

So I toned it down a bit. I hope it is clear enough now :-/

akesandgren reviewed (approved):
Still LGTM

akesandgren commented:
Going in, thanks @Flamefire!
